Add classic "Welcome" intro

MiraiSubject opened the pull request:
This adds an intro screen to osu! lazer with the soundtrack from older osu! stable versions (also known as fallback), nekodex - welcome to osu!

- [x] Depends on https://github.com/ppy/osu-resources/pull/102
- [x] Depends on https://github.com/ppy/osu-resources/pull/103
- [x] Depends on https://github.com/ppy/osu/pull/9236
- [x] Depends on ppy/osu-framework#3648

MiraiSubject committed:
Add fallback intro screen

peppy commented on osu.Game/Configuration/IntroSequence.cs:
```diff
@@ -6,6 +6,7 @@ namespace osu.Game.Configuration
     public enum IntroSequence
     {
         Circles,
+        Fallback,
```
Let's call it "Welcome" for now.

peppy commented:
If you're going to implement this, please make the welcome/seeya sound also match. also needs the piano reverse-reverb. All resources attached.

[resources.zip](https://github.com/ppy/osu/files/4715337/resources.zip)

peppy reviewed (changes requested):
As specified

MiraiSubject committed:
Implement welcome and seeya samples

MiraiSubject committed:
Change "Fallback" to "Welcome" visually

MiraiSubject committed:
Implement basic intro sequence

MiraiSubject committed:
Rename IntroFallback classes to IntroLegacy

This commit also renames files accordingly with https://github.com/ppy/osu-resources/pull/103

MiraiSubject committed:
Merge branch 'master' into fallback-intro

MiraiSubject commented:
I added the welcome and seeya sounds that match this sequence to the screens and an initial implementation of the legacy start up animation. 

Right now I'm having some trouble restoring the glow using `GlowingSpriteText` so as an alternative for now an `OsuSpriteText` is used. 

The visualizer, shown below in fallback, is also still absent. I'm still thinking of a way on how to properly approach the implementation of that. 

![osu!_2020-06-02_13-18-24](https://user-images.githubusercontent.com/803255/83514292-add9ac80-a4d3-11ea-84f9-9ca658f864e4.png)

peppy commented:
Which glow are you talking about? It should be baked into the sprite.

I'll check the stable implementation if you want the exact transforms.

MiraiSubject commented:
I can use the original png sprite. In that case the original transforms would definitely be desired and make implementing it easier.

I had the thought that using the built-in Drawable for `OsuSpriteText` was preferred.

MiraiSubject committed:
Rename legacy to welcome to match osu-resources

peppy commented:
Definitely use the sprite. I don't think the correct font is even available.

MiraiSubject committed:
Add original sprite and visualiser

Notes: This is using a modified version of welcome.osz to facilitate the visualiser and the animation of the sprite is not accurate.

MiraiSubject commented:
I mentioned on the discord that right now the piano reverb is a separate mp3, but the `LogoVisualisation` does not take arbitrary audio streams. 

So right now I've prepended the piano reverb to the audio file inside `welcome.osz` to achieve the desired start up effect. 

Because I'm basically editing the original beatmap, I'd like to know if this is the desired direction or if I should approach it differently.

MiraiSubject commented on osu.Game/Screens/Menu/IntroWelcome.cs:
```diff
@@ -0,0 +1,152 @@
+                                        Anchor = Anchor.Centre,
+                                        Origin = Anchor.Centre,
+                                        Alpha = 0.5f,
+                                        AccentColour = Color4.Blue,
```
This does nothing at the moment. The accent color gets overriden in `LogoVisualisation`

MiraiSubject committed:
Modify LogoVisualisation to allow color changes

Also change the color from blue to dark blue

MiraiSubject committed:
Fix styling error

peppy commented:
You could just playback the piano reverb manually. As a sample, similar to the welcome voice line.

MiraiSubject committed:
Revert piano reverb to a separate sample

MiraiSubject commented:
Alright, I have reverted the sample back!

> I'll check the stable implementation if you want the exact transforms.

May I have the original transforms of the `Welcome` sprite ?

MiraiSubject committed:
Change transforms to roughly match fallback visually

MiraiSubject committed:
Code formatting fixed

Somehow slipped through after pushing

MiraiSubject commented:
I've managed to make a rough approximation of the transforms by recording a video of the fallback intro and then by playing it back at a slower rate I was able to observe the individual animations happening on the `Welcome` sprite.

bdach commented on osu.Game/Screens/Menu/IntroWelcome.cs:
```diff
@@ -0,0 +1,163 @@
+                    welcomeText.FadeIn(remainingTime());
+                    welcomeText.ScaleTo(welcomeText.Scale + new Vector2(0.1f), remainingTime(), Easing.Out).OnComplete(_ =>
+                    {
+                        elementContainer.Remove(visualizer);
+                        circleContainer.Remove(blackCircle);
+                        elementContainer.Remove(circleContainer);
+                        Remove(welcomeText);
+                        visualizer.Dispose();
+                        blackCircle.Dispose();
+                        welcomeText.Dispose();
```
I'm not overly familiar with drawable lifetime stuff, but isn't `Expire()` what you actually want here? If you expire the elements they should automatically get cleaned up from the containers. [There's even a framework-side test case for that.](https://github.com/ppy/osu-framework/blob/2637ba3af75e5ef0d3e05388309e153e457f489c/osu.Framework.Tests/Containers/TestSceneContainerState.cs#L222-L239)

bdach commented on osu.Game/Screens/Menu/LogoVisualisation.cs:
```diff
@@ -70,6 +70,7 @@ public class LogoVisualisation : Drawable, IHasAccentColour
 
         private IShader shader;
         private readonly Texture texture;
+        public bool IsIntro = false;
```
Flags added for specific purposes like this are always asking for trouble. It's doing things backwards; to reuse this visualisation for the legacy intro we should look for ways to *remove* the user colour changing logic for that particular use case entirely, and to keep it in all the other places that need it right now, and not ways to override it.

I can't say I see a very good way of achieving this right now, but I think extracting out a simplified variant of `LogoVisualisation` that doesn't have `User` or `Skin`, and then inheriting from it to add that functionality back where it's needed right now could be a decent idea to start with. That might end up being a separate prerequisite PR, however, depending if any renames/significant code reshuffle happens.

MiraiSubject replied:
I've implemented this in https://github.com/ppy/osu/pull/9236. The naming for the new base class for `LogoVisualisation` might be a bit questionable though.

bdach commented on osu.Game/Screens/Menu/IntroWelcome.cs:
```diff
@@ -0,0 +1,163 @@
+                welcome?.Play();
+                pianoReverb?.Play();
+
+                Scheduler.AddDelayed(delegate
```
`delegate` is a really old thing now. Lambdas are a bit more hip and modern nowadays.

```suggestion
                Scheduler.AddDelayed(() =>
```

MiraiSubject replied:
Yeah you're right! That was just a blind copy-paste from `IntroCircles` really. I'll include this with the change for the drawable lifetimes.

bdach reviewed (dismissed):
Haven't ran this yet, just skimmed the source.

MiraiSubject committed:
Implement suggested changes

Note: LogoVisualisation is likely going to be needed in a separate PR to conform to the review.

MiraiSubject committed:
Removed logovisualisation changes

Now depends on https://github.com/ppy/osu/pull/9236 for accent color changes to apply

MiraiSubject committed:
Removed unneeded properties

peppy committed:
Merge branch 'master' into fallback-intro

MiraiSubject committed:
Change hash to be lowercase and change sample directories

peppy commented on osu.Game/Screens/Menu/IntroWelcome.cs:
```diff
@@ -0,0 +1,149 @@
+        [BackgroundDependencyLoader]
+        private void load(AudioManager audio)
+        {
+            Seeya = audio.Samples.Get(@"Intro/welcome/seeya");
```
unnecessarily loading two samples (the base is also loading). revert the `SampleChannel` to private and make a `virtual string SeeyaSampleName => "Intro/seeya"`

peppy commented on osu.Game/Screens/Menu/IntroWelcome.cs:
```diff
@@ -0,0 +1,149 @@
+            if (MenuVoice.Value)
+            {
+                welcome = audio.Samples.Get(@"Intro/welcome/welcome");
+                pianoReverb = audio.Samples.Get(@"Intro/welcome/welcome_piano");
```
piano revert plays even when menuvoice is off

peppy commented on osu.Game/Screens/Menu/IntroWelcome.cs:
```diff
@@ -0,0 +1,149 @@
+                        Width = 750,
+                        Height = 78,
+                        Alpha = 0,
+                        Texture = textures.Get(@"Welcome/welcome_text@2x")
```
`@2x` shouldn't be here. everything in lazer is @2x.

MiraiSubject replied:
The texture doesn't display at all when I remove `@2x`, I'd have assumed first as well that it wouldn't be needed, but it's actually not visible on my side. 

`@2x` removed, so `textures.Get(@"Welcome/welcome_text")`
![dotnet_2020-06-13_16-28-06](https://user-images.githubusercontent.com/803255/84571342-e78f9a80-ad92-11ea-869e-2a4a0c367067.png)

`@2x` added
![dotnet_2020-06-13_16-27-39](https://user-images.githubusercontent.com/803255/84571340-e65e6d80-ad92-11ea-9eab-c209fd4fa4c0.png)

peppy commented on osu.Game/Screens/Menu/IntroWelcome.cs:
```diff
@@ -0,0 +1,149 @@
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        Scale = new Vector2(0.3f),
+                        Width = 750,
```
width and height specification should not be required. it's automatic when you set a texture.

MiraiSubject replied:
```csharp
Width = 750,
Height = 78,
```
This is hardcoded like this to prevent the game from proportionally scaling it. The original dimensions for the file are `750*156`

`welcomeText.ResizeHeightTo(welcomeText.Height * 2, 500, Easing.In);` resizes the height from `750*78` to `750*156`

peppy commented on osu.Game/Screens/Menu/IntroWelcome.cs:
```diff
@@ -0,0 +1,149 @@
+            {
+                base.LoadComplete();
+
+                double remainingTime() => delay_step_two - TransformDelay;
```
shouldn't be a function. also not sure this is correct.

MiraiSubject committed:
Implement changes from review

Moves seeya back to the introscreen and uses a virtual string to change whenever it's needed and removed remainingTime()

MiraiSubject committed:
Merge branch 'master' of https://github.com/ppy/osu into fallback-intro

MiraiSubject committed:
Use RelativeSizeAxes for width

peppy committed:
Merge branch 'master' into fallback-intro

peppy committed:
Update resources

peppy committed:
Update welcome text sprite location

peppy committed:
Fix animations and general code quality

peppy committed:
Merge branch 'master' into fallback-intro

peppy commented:
I've tidied things up. The main issue remaining is that the menu visualisation does not correctly animate due to not receiving the amplitudes of the piano reverb sample. I'll look into this further - a fix will either require some changes to the sample playback logic or combination of the sample into the track (may actually be the simplest way for now).

peppy committed:
Fix samples being played too early

peppy committed:
Update naming

peppy committed:
Read amplitudes from piano reverb source

peppy commented:
I've applied the final touches to this, including reading the amplitudes from the reverb sample (requires a framework change).

smoogipoo commented:
The new amplitude source stuff crashes when changing rulesets in song select:

```
[runtime] 2020-06-23 07:21:44 [error]: System.NullReferenceException: Object reference not set to an instance of an object.
[runtime] 2020-06-23 07:21:44 [error]: at osu.Game.Screens.Menu.LogoVisualisation.addAmplitudesFromSource(IHasAmplitudes source) in /home/smgi/Repos/osu/osu.Game/Screens/Menu/LogoVisualisation.cs:line 161
[runtime] 2020-06-23 07:21:44 [error]: at osu.Game.Screens.Menu.LogoVisualisation.updateAmplitudes() in /home/smgi/Repos/osu/osu.Game/Screens/Menu/LogoVisualisation.cs:line 113
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal() in /home/smgi/Repos/osu-framework/osu.Framework/Threading/ScheduledDelegate.cs:line 88
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Threading.Scheduler.Update() in /home/smgi/Repos/osu-framework/osu.Framework/Threading/Scheduler.cs:line 102
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Drawable.UpdateSubTree() in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Drawable.cs:line 477
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.updateChild(Drawable c) in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 883
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree() in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 864
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.updateChild(Drawable c) in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 883
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree() in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 864
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.updateChild(Drawable c) in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 883
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree() in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 864
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.updateChild(Drawable c) in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 883
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree() in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 864
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.updateChild(Drawable c) in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 883
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree() in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 864
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.updateChild(Drawable c) in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 883
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree() in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 864
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.updateChild(Drawable c) in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 883
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree() in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 864
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.updateChild(Drawable c) in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 883
[runtime] 2020-06-23 07:21:44 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree() in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable.cs:line 864
```

peppy commented:
ah yes, because they can still be null in some cases for whatever reason. shall we fix that framework-side by initialising the amplitude array to an empty array, or leave as nullable?

smoogipoo commented:
I'd probably prefer initialising to an empty array...

peppy committed:
Merge branch 'master' into fallback-intro

peppy committed:
Fix background appearing too late

smoogipoo reviewed (approved):
Seems good
